File: packages/domain-contract/src/entity-preview.test.ts
import { describe, expect, it } from "bun:test"
import {
  CHIP_LABEL_MAX,
  ENTITY_REF_ICON,
  ENTITY_REF_KINDS,
  entityChipDetail,
  entityChipLabel,
  entityKindNoun,
  entityRefIcon,
  groupPreviewRefs,
  type EntityRefLike,
} from "./entity-preview"
import fixture from "../fixtures/entity-chip.json" with { type: "json" }
import icons from "../../icons/icons.json" with { type: "json" }

// EXP-920: the entity-preview chip rule, locked ×4 (desktop
// domain::entity_preview, iOS EntityPreviewTests, Android EntityPreviewTest)
// against the ONE contract fixture — same cases, same test names.

interface ChipCase {
  name: string
  ref: EntityRefLike
  label: string
  detail: string | null
  icon: string
}

interface GroupCase {
  name: string
  refs: EntityRefLike[]
  groups: { ref: number; members: number[] }[]
}

const chips = fixture.chips as ChipCase[]
const groups = fixture.groups as GroupCase[]

describe(`entity preview`, () => {
  it(`every chip case renders byte-exact`, () => {
    expect(chips.length).toBeGreaterThanOrEqual(24)
    for (const { name, ref, label, detail, icon } of chips) {
      expect(entityChipLabel(ref), name).toBe(label)
      expect(entityChipDetail(ref), name).toBe(detail)
      expect(entityRefIcon(ref), name).toBe(icon)
    }
  })

  it(`every group case groups the same`, () => {
    expect(groups.length).toBeGreaterThanOrEqual(6)
    for (const { name, refs, groups: expected } of groups) {
      const actual = groupPreviewRefs(refs).map((group) => ({
        ref: refs.indexOf(group.ref),
        members: group.members.map((member) => refs.indexOf(member)),
      }))
      expect(actual, name).toEqual(expected)
    }
  })

  it(`every contract kind has an icon concept that exists, and a noun`, () => {
    const semantic = (icons as { semantic: Record<string, string> }).semantic
    for (const kind of ENTITY_REF_KINDS) {
      expect(ENTITY_REF_ICON[kind], kind).toBeDefined()
      expect(semantic[ENTITY_REF_ICON[kind]!], kind).toBeDefined()
      expect(entityKindNoun(kind, 1), kind).not.toBe(`item`)
      expect(entityKindNoun(kind, 2), kind).not.toBe(`items`)
    }
    expect(Object.keys(ENTITY_REF_ICON).sort()).toEqual([...ENTITY_REF_KINDS].sort())
  })

  it(`the fixture covers every kind's chip`, () => {
    const seen = new Set(chips.map((c) => (c.ref.kind === `list` ? `list` : c.ref.kind)))
    for (const kind of ENTITY_REF_KINDS) expect(seen.has(kind), kind).toBe(true)
  })

  it(`a label never exceeds the cap`, () => {
    for (const { ref } of chips) {
      expect(Array.from(entityChipLabel(ref)).length).toBeLessThanOrEqual(CHIP_LABEL_MAX)
    }
  })
})

File: apps/android/app/src/test/java/com/exponential/app/domain/EntityPreviewTest.kt
package com.exponential.app.domain

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * EXP-920: the entity-preview chip rule, locked ×4 (web `entity-preview.test.ts`,
 * iOS `EntityPreviewTests`, desktop `domain::entity_preview`) against the ONE
 * contract fixture `entity-chip.json` — same cases, same test names.
 */
class EntityPreviewTest {

    private val fixture = Json.parseToJsonElement(contractFixtureJson("entity-chip.json")).jsonObject

    @Test
    fun everyChipCaseRendersByteExact() {
        val cases = fixture.getValue("chips").jsonArray
        assertTrue(cases.isNotEmpty())
        cases.forEach { element ->
            val case = element.jsonObject
            val name = case.getValue("name").jsonPrimitive.content
            // The fixture's refs go in UNPARSED, the way web feeds them to the
            // rule: the padding/unknown-kind cases exercise the rule's own
            // trimming and fallbacks, not the parser's.
            val ref = rawRef(case.getValue("ref").jsonObject)
            assertEquals(name, case.getValue("label").jsonPrimitive.content, EntityPreview.chipLabel(ref))
            val detail = case.getValue("detail").let { if (it is JsonNull) null else it.jsonPrimitive.content }
            assertEquals(name, detail, EntityPreview.chipDetail(ref))
            assertEquals(name, case.getValue("icon").jsonPrimitive.content, EntityPreview.refIcon(ref))
        }
    }

    @Test
    fun everyGroupCaseGroupsTheSame() {
        val cases = fixture.getValue("groups").jsonArray
        assertTrue(cases.isNotEmpty())
        cases.forEach { element ->
            val case = element.jsonObject
            val name = case.getValue("name").jsonPrimitive.content
            val refs = case.getValue("refs").jsonArray.map { rawRef(it.jsonObject) }
            val expected = case.getValue("groups").jsonArray.map { group ->
                val obj = group.jsonObject
                EntityPreview.Group(
                    ref = refs[obj.getValue("ref").jsonPrimitive.int],
                    members = obj.getValue("members").jsonArray.map { refs[it.jsonPrimitive.int] },
                )
            }
            assertEquals(name, expected, EntityPreview.groupRefs(refs))
        }
    }

    @Test
    fun `every contract kind has an icon concept`() {
        DomainContract.entityRefKindValues.forEach { kind ->
            assertNotNull("icon for $kind", EntityPreview.ICON[kind])
        }
        assertEquals(DomainContract.entityRefKindValues.toSet(), EntityPreview.ICON.keys)
    }

    @Test
    fun `the parser mirrors the web one`() {
        // An unknown kind drops; a blank id drops; strings trim + clamp; count rounds.
        assertNull(EntityRef.parse(Json.parseToJsonElement("""{"kind":"thing","id":"x"}""")))
        assertNull(EntityRef.parse(Json.parseToJsonElement("""{"kind":"issue","id":"  "}""")))
        assertNull(EntityRef.parse(Json.parseToJsonElement("""{"kind":"issue"}""")))
        assertNull(EntityRef.parse(Json.parseToJsonElement("""["issue"]""")))
        val long = "x".repeat(DomainContract.expToolPreviewTextMax + 20)
        val parsed = EntityRef.parse(
            Json.parseToJsonElement(
                """{"kind":"issue","id":" i-1 ","identifier":" EXP-1 ","title":"$long","count":2.6}""",
            ),
        )!!
        assertEquals("i-1", parsed.id)
        assertEquals("EXP-1", parsed.identifier)
        assertEquals(DomainContract.expToolPreviewTextMax, parsed.title!!.length)
        assertEquals(3, parsed.count)
        // A blank title is absent, a negative count is absent, a string count is absent.
        val sparse = EntityRef.parse(
            Json.parseToJsonElement("""{"kind":"board","id":"b","title":"  ","count":-1}"""),
        )!!
        assertNull(sparse.title)
        assertNull(sparse.count)
        assertNull(EntityRef.parse(Json.parseToJsonElement("""{"kind":"list","id":"issue","count":"3"}"""))!!.count)
    }

    /** A fixture ref as written, parser untouched. */
    private fun rawRef(obj: kotlinx.serialization.json.JsonObject): EntityRef = EntityRef(
        kind = obj.getValue("kind").jsonPrimitive.content,
        id = obj.getValue("id").jsonPrimitive.content,
        identifier = obj["identifier"]?.jsonPrimitive?.content,
        title = obj["title"]?.jsonPrimitive?.content,
        count = obj["count"]?.jsonPrimitive?.int,
    )
}
